File: Algorithms/Medium/ValidateStackSequences.go
package medium

// Time complexity: O(n)
// Space complexity: O(n)
func validateStackSequences(pushed []int, popped []int) bool {
	stack := make([]int, 0, len(pushed))
	i, j := 0, 0

	push := func(item int) {
		stack = append(stack, item)
	}
	pop := func() int {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		return top
	}

	for i < len(pushed) {
		if len(stack) > 0 && stack[len(stack)-1] == popped[j] {
			pop()
			j++
		} else {
			push(pushed[i])
			i++
		}
	}

	for j < len(popped) {
		if pop() != popped[j] {
			return false
		}
		j++
	}

	return true
}

File: Algorithms/Medium/946.Validate-Stack-Sequences/ValidateStackSequences.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def validateStackSequences(self, pushed: list[int], popped: list[int]) -> bool:
        stack = []
        i, j = 0, 0

        while i < len(pushed):
            if stack and stack[-1] == popped[j]:
                stack.pop()
                j += 1
            else:
                stack.append(pushed[i])
                i += 1

        while j < len(popped):
            if stack.pop() != popped[j]:
                return False
            j += 1

        return True
